File: tasks/oj/boj/python/21/21922.py
## 시간초과: 디버그 코드로 m, n이 큰 수인 연구실 맵 생성 후 추가... 
## 오래 걸림
from sys import stdin
lab = []
conditioners = []
airpath = set()

n, m = map(int, stdin.readline().split())
for y in range(n):
    arr = list(map(int, stdin.readline().split()))
    if 9 in arr:
        for x in range(m):
            if arr[x] == 9:
                conditioners += [(y, x)]
    lab += [arr]

for conditioner in conditioners:
    for direction in range(4): # up right down left
        f = conditioner # focus
        back = False
        if direction == 0:
            offset = [-1, 0]
        elif direction == 1:
            offset = [0, 1]
        elif direction == 2:
            offset = [1, 0]
        else:
            offset = [0, -1]
        d = direction
        while not back:
            airpath.add(f)
            ns = (f[0]+offset[0], f[1]+offset[1]) # next_step
            if ns[0] >= 0 and ns[1] >= 0 and ns[0] < n and ns[1] < m:
                if lab[ns[0]][ns[1]] == 0:
                    pass
                elif lab[ns[0]][ns[1]] == 1:
                    if d == 1 or d == 3:
                        airpath.add(ns)
                        back = True
                elif lab[ns[0]][ns[1]] == 2:
                    if d == 0 or d == 2:
                        airpath.add(ns)
                        back = True
                elif lab[ns[0]][ns[1]] == 3:
                    if d == 0:
                        d = 1
                        offset = [0, 1]
                    elif d == 1:
                        d = 0
                        offset = [-1, 0]
                    elif d == 2:
                        d = 3
                        offset = [0, -1]
                    elif d == 3:
                        d = 2
                        offset = [1, 0]
                elif lab[ns[0]][ns[1]] == 4:
                    if d == 0:
                        d = 3
                        offset = [0, -1]
                    elif d == 1:
                        d = 2
                        offset = [1, 0]
                    elif d == 2:
                        d = 1
                        offset = [-1, 0]
                    elif d == 3:
                        d = 0
                        offset = [0, 1]
                else:
                    airpath.add(ns)
                    back = True
                f = ns
            else:
                back = True
print(len(airpath))

File: tasks/oj/boj/rust/21922.rs
use std::{io, collections::VecDeque};

fn main() {
  let mut puts = String::new();
  io::stdin().read_line(&mut puts).ok();
  let (n, m) = {
    let split: Vec<i32> = puts.split_whitespace()
      .map(|x| x.parse::<i32>().unwrap())
      .collect();
    (split[0], split[1])
  };
  let mut map: Vec<Vec<i32>> = vec![vec![-1; m as usize]; n as usize];
  let mut aircons: Vec<(i32, i32)> = Vec::new();
  for i in 0..n {
    puts.clear();
    io::stdin().read_line(&mut puts).ok();
    let row: Vec<i32> = puts.split_whitespace()
      .map(|x| x.parse::<i32>().unwrap())
      .collect();
    for j in 0..m {
      if row[j as usize] == 9 {
        aircons.append(&mut vec![(i, j)]);
      }
    }
    map[i as usize] = row;
  }
  let mut res: Vec<Vec<i32>> = vec![vec![0; m as usize]; n as usize];
  let mut queue: VecDeque<(i32, i32, i32)> = VecDeque::new(); // i, j, direction
  // direction: 0 up / 1 left / 2 down / 3 right
  for aircon in aircons {
    queue.push_back((aircon.0 - 1, aircon.1, 0));
    queue.push_back((aircon.0 + 1, aircon.1, 2));
    queue.push_back((aircon.0, aircon.1 - 1, 1));
    queue.push_back((aircon.0, aircon.1 + 1, 3));
    res[aircon.0 as usize][aircon.1 as usize] = 1;
  }
  while !queue.is_empty() {
    let now = queue.pop_front().unwrap();
    if now.0 < 0 || now.0 >= n || now.1 < 0 || now.1 >= m {
      continue;
    }
    res[now.0 as usize][now.1 as usize] = 1;
    let tile = map[now.0 as usize][now.1 as usize];
    if tile == 0 {
      if now.2 == 0 { queue.push_back((now.0-1, now.1, 0)); }
      else if now.2 == 1 { queue.push_back((now.0, now.1-1, 1)); }
      else if now.2 == 2 { queue.push_back((now.0+1, now.1, 2)); }
      else { queue.push_back((now.0, now.1+1, 3)); }
    } else if tile == 1 {
      if now.2 == 0 { queue.push_back((now.0-1, now.1, 0)); } // up
      else if now.2 == 2 { queue.push_back((now.0+1, now.1, 2)); } // down
      // horizontal will be ignored
    } else if tile == 2 {
      if now.2 == 1 { queue.push_back((now.0, now.1-1, 1)); } // left
      else if now.2 == 3 { queue.push_back((now.0, now.1+1, 3)); } // right
      // vertical will be ignored
    } else if tile == 3 {
      if now.2 == 0 { queue.push_back((now.0, now.1+1, 3)); } // up to right
      else if now.2 == 1 { queue.push_back((now.0+1, now.1, 2)); } // left to down
      else if now.2 == 2 { queue.push_back((now.0, now.1-1, 1)); } // down to left
      else { queue.push_back((now.0-1, now.1, 0)); } // right to up
    } else if tile == 4 {
      if now.2 == 0 { queue.push_back((now.0, now.1-1, 1)) } // up to left
      else if now.2 == 1 { queue.push_back((now.0-1, now.1, 0)) } // left to up
      else if now.2 == 2 { queue.push_back((now.0, now.1+1, 3)) } // down to right
      else { queue.push_back((now.0+1, now.1, 2)) } // right to down
    }
  }
  let mut sum: i32 = 0;
  for arr in res {
    sum += arr.iter().sum::<i32>();
  }
  println!("{}", sum);
}